src/pixelator/logging.py

```diff
@@ -192,33 +195,43 @@ def close(self):
         """Close down the logging setup."""
         self._shutdown_listener()
 
-    def __exit__(self, exc_type, exc_value, traceback):
+    def __exit__(self, exc_type, exc_value, traceback_obj):
         """Exit the context manager.
 
         This will shut down the logging process if needed.
         """
-        self.close()
-        # Reraise exception higher up the stack
-        return False
-
 
-def handle_unhandled_exception(exc_type, exc_value, exc_traceback):
-    """Handle "unhandled" exceptions."""
-    pixelator_root_logger = logging.getLogger()
+        def log_exception(exc_type, exc_value, traceback_obj):
+            if issubclass(exc_type, click.exceptions.Exit):
+                # click will raise a click.exceptions.Exit exception
+                # when exiting the application, and we don't want to
+                # put that as an error in the log.
+                return
 
-    if issubclass(exc_type, KeyboardInterrupt):
-        # Will call default excepthook
-        sys.__excepthook__(exc_type, exc_value, exc_traceback)
-        return
-
-    # Create a critical level log message with info from the except hook.
-    pixelator_root_logger.critical(
-        "Unhandled exception", exc_info=(exc_type, exc_value, exc_traceback)
-    )
+            self._root_logger.critical(
+                "Unhandled exception of type: {}".format(exc_type.__name__)
+            )
+            self._root_logger.critical("Exception message was: {}".format(exc_value))
+            for item in traceback.format_exception(exc_type, exc_value, traceback_obj):
+                for line in item.splitlines():
+                    self._root_logger.critical(line)
 
+        try:
+            # It seems that sometimes the current exception is not passed
+            # to the __exit__ method. This allows us to check if there
+            # is any raised exception and make sure this gets added to
+            # the log.
+            if exc_type is None:
+                exc_type, exc_value, traceback_obj = sys.exc_info()
+                if exc_type is not None:
+                    log_exception(exc_type, exc_value, traceback_obj)
+            else:
+                log_exception(exc_type, exc_value, traceback_obj)
+        finally:
+            self.close()
 
-# Assign the excepthook to the handler
-sys.excepthook = handle_unhandled_exception
+        # Reraise exception higher up the stack
+        return False
```

Comment:
No, which is a bit confusing. But exit should return `True` if it suppresses the exception, otherwise it should return `False`. https://docs.python.org/3/reference/datamodel.html#object.__exit__